games/tic_tac_toe/ultimate_tic_tac_toe/game.py

```diff
@@ -0,0 +1,89 @@
+from itertools import product
+from typing import Callable, List, overload
+
+import numpy as np
+from base.game import Game
+from games.tic_tac_toe.tic_tac_toe_block_logic import (
+    EMPTY_SIGN, TicTacToeBlockLogic, tic_tac_toe_block_logic_factory)
+from games.tic_tac_toe.ultimate_tic_tac_toe.action import UTTTAction
+from games.tic_tac_toe.ultimate_tic_tac_toe.state import UTTTState
+from numpy.typing import NDArray
+from PIL import Image
+
+
+class UltimateTicTacToeGame(Game[UTTTState, UTTTAction]):
+
+    def __init__(
+        self, player_sign='X', opponent_sign='O',
+        block_logic_factory: Callable[[str, str],
+                                      TicTacToeBlockLogic] = tic_tac_toe_block_logic_factory):
+        self.player_sign = player_sign
+        self.opponent_sign = opponent_sign
+        self.block_logic = block_logic_factory(player_sign, opponent_sign)
+        initial_state = self._initial_game_state()
+        super().__init__(initial_state)
+
+    def _initial_game_state(self) -> UTTTState:
+        return UTTTState(
+            board=np.full((self.blocks_count,
+                           *self.block_shape),
+                          EMPTY_SIGN),
+            curr_block_idx=4)
+
+    @property
+    def block_shape(self):
+        return self.block_logic.block_shape()
+
+    def actions_for(self, state: UTTTState, is_opponent: bool) -> List[UTTTAction]:
+        sign = self._get_sign(is_opponent)
+
+        row_count, col_count = self.block_shape
+
+        def positions(block_idxs):
+            return((block_idx, row, col) for block_idx, row, col in product(block_idxs, range(row_count), range(col_count))
+                   if state.board[block_idx, row, col] == EMPTY_SIGN and not self.is_block_terminated(state.board[block_idx]))
+
+        valid_blocks_idxs = range(self.blocks_count) if self.is_block_terminated(
+            state.board[state.curr_block_idx]) else [state.curr_block_idx]
+
+        return [UTTTAction(sign, block, row, col) for block, row, col in positions(valid_blocks_idxs)]
+
+    @property
+    def blocks_count(self):
+        return 9
```

Comment:
Z ciekawości, czemu zrobiłeś to jako `@property` a nie `self.blocks_count = 9`?

Reply:
Głównie dlatego, że wcześnie to była funkcja która dostawała `board` i zwracała `shape[0]`. A potem dlaczegoś zmieniłem podejście i usunąłem `board` z argumentów.
A potem uznałem, że skoro to jest funkcja bez argumentów, to może warto już zrobić to jako property 😄 

Zmieniłem na `self.blocks_count` 😉